### PR TITLE
[Issue 15973]: nextPow2 and truncPow2 rely on processor specific …

### PR DESCRIPTION
**Pull request by JackStouffer**
…behavior

### CODE REVIEW

**Comment by JackStouffer** [comment]
This is a blocking issue for LDC, https://github.com/dlang/phobos/pull/3755/files#r61540356, so please review and pull as soon as possible.

File: std/math.d [hunk 1]
```diff
@@ -7341,6 +7341,8 @@ private T powIntegralImpl(PowType type, T)(T val)
 
     if (x == 0)
         return x;
+    if (x == Unqual!T.max || x == Unqual!T.min)
```
Comment (rainers) [inline std/math.d]: Thanks for the quick action.

I think the condition must only apply to PowType.ceil and should distinguish between signed and unsigned:
```
static if (type == PowType.ceil)
{
    if(x > T.max / 2 || x == T.min)
        return 0;
}
```

File: std/math.d [hunk 2]
```diff
@@ -7525,15 +7527,15 @@ T truncPow2(T)(const T val) if (isFloatingPoint!T)
     assert(truncPow2(-5) == -4);
     assert(truncPow2(-20) == -16);
 
-    assert(truncPow2(uint.max) == int.max + 1);
+    assert(truncPow2(uint.max) == 0);
```
Comment (rainers) [inline std/math.d]: truncPow2 was fine, no need to change it.

**Comment by JackStouffer** [comment]
@rainers fixed

**Comment by rainers** [comment]
Looks good. For an additional bonus
- add some tests for the new border cases:
```
      assert(nextPow2(uint.max/2) == uint.max/2 + 1);
      assert(nextPow2(uint.max/2 + 2) == 0);
      assert(nextPow2(int.max/2) == int.max/2 + 1);
      assert(nextPow2(int.max/2 + 2) == int.max/2 + 1);
      assert(nextPow2(int.min + 1) == int.min);
```
- document overflow/underflow behaviour

**Comment by wilzbach** [comment]
Just to throw this into the discussion: http://en.cppreference.com/w/cpp/numeric/math/pow

> If the base is finite negative and the exponent is finite but not an integer value, it causes a domain error.
> If both base and exponent are zero, it may also cause a domain error on certain implementations.
> If base is zero and exponent is negative, it may cause a domain error or a pole error (or none, depending on the library implementation).
> The function may also cause a range error if the result is too great or too small to be represented by a value of the return type.

**Comment by rainers** [comment]
> Just returning 0 seems a bit too arbitrary too me :(

My impression is that `nogc nothrow` is natural for this function, so throwing an exception/error (D's version of domain/range error) might be inappropriate. For the float versions returning NaN could be an option, but that doesn't exist for integers. 0 is a value that can never be returned for values that don't overflow.

**Comment by JackStouffer** [comment]
@rainers shouldn't it be

```
    assert(nextPow2(int.max / 2 + 2) == 0);
    assert(nextPow2(int.min + 1) == 0);
```

**Comment by rainers** [comment]
> assert(nextPow2(int.max / 2 + 2) == 0); 

Yes.

> assert(nextPow2(int.min + 1) == 0);

No. `nextPow2(-0x7fff_ffff) == -0x8000_0000`

**Comment by rainers** [comment]
I guess the additional checks need to be before applying `abs` (which could also just be `-x`)

**Comment by JackStouffer** [comment]
@rainers Fixed

**Comment by rainers** [comment]
Thanks.
Sorry to be such a nuisance but I noticed more:
- dmd 2.071 is unable to inline the code, so it fails with option `-inline`. This version seems to work:
```
pragma(inline, true)
private T powIntegralImpl(PowType type, T)(T val)
{
    import core.bitop : bsr;

    if (val == 0 || (type == PowType.ceil && (val > T.max / 2 || val == T.min)))
        return 0;
    else static if (isSigned!T)
        return val < 0 ? -(T(1) << bsr(-val) + type) : T(1) << bsr(val) + type;
    else
        return (T(1) << bsr(val) + type);
}
```
You have to ensure that there is not more than a single if/else expression in the function. Please note that the unittests are built without inlining enabled. 
I expect the float version isn't very fast anyway, so I'd rather remove the `pragma(inline.true)` there.
* The float version also returns 0 for overflow, so there is no need to add the conditional in the documentation (+/-infinity might also be a good choice for overflowing floats, though)
* some versions of LLVM also stumble for `assert(truncPow2(int.min) == int.min)`. Unfortunately I cannot reproduce on Windows, but I suspect it originates in `-int.min == int.min`. I wonder what use case exists for nextPow2 for negative numbers...

**Comment by rainers** [comment]
> some versions of LLVM also stumble for assert(truncPow2(int.min) == int.min)

I think I understand this now: `bsr(int.min) == 63`, so `(1 << 63) == int.min` fails without the truncation of the shift.

**Comment by JackStouffer** [comment]
@rainers Sorry it took me a while to update this, but I added your fixes

**Comment by rainers** [comment]
Auto-merge toggled on

**Comment by rainers** [comment]
> @rainers Sorry it took me a while to update this, but I added your fixes

Thanks. truncPow2(int.min) works as expected for dmd, though might need a change for other platforms later.

**Pull request merged**